### DISCUSSION TITLE
Can vagrant libvirt work without a libvirt daemon running all the time?

### DISCUSSION THREAD

**Discussion by imperialguy** [vagrant-libvirt/vagrant-libvirt#1192] [Q&A]
Can vagrant libvirt plugin work without a libvirt daemon running all the time?

**Comment by electrofelix** [comment]
Not really sure what your asking about? If you stop the libvirt service by default it'll shutdown any VMs started by it. Maybe that has been split out these days, not at a machine to check. If the VMs started via vagrant-libvirt have been shutdown not sure how useful the plugin can be at that point. What is it you're hoping to do?

**Comment by imperialguy** [comment]
I mean can the plugin itself control the `libvirt` service - bring up the service when it's spinning any VMs and bring it down when there's no running VMs. For e.g. vagrant creates a `VBoxSVC --auto-shutdown` and `VBoxheadless` processes (on Mac) automatically when vagrant spins up VMs using Virtual Box and then kills that process when VMs are suspended. So, perhaps the plugin can spin up the libvirt service automatically or may be not use the `libvirt` daemon, instead use the utilities provided by `libvirt` (`virtsh`, `virt-qemu-run`, `virt-admin` etc..) to perform whatever operations vagrant is trying to do.

**Reply by electrofelix** [reply]
No it's not currently supported. A quick read suggests the standalone method should be possible, though what the networking would need isn't clear, so that's possibly the main tricky point along with managing tracking the machine id so it's known when it's running, and finally ensuring it's started in a way that prevents termination when the parent vagrant process exits (already done for port forwarding via SSH so just copy that).

I'd suspect that it might be a little beyond what I can put effort towards for a while as there is a few things I need to get to first, however I don't mind providing a few pointers if your interested in implementing.

The main thing would be to define this as a specific driver type that can be selected. I did move some code into a driver.rb file with the thought that at some stage if there was a need to support different behaviour it might be best abstracted in one place. It's incomplete though, so some refactoring might be needed to make it easy to switch to this standalone behaviour

**Comment by imperialguy** [comment]
Separate driver sounds good. I am getting this error on Mac, no matter which libvirt vagrant box I try to run. 

```
/Users/vkarri/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.3.0/lib/vagrant-libvirt/driver.rb:59:in `open_read_only': Call to virConnectOpenReadOnly failed: Failed to connect socket to '/usr/local/var/run/libvirt/libvirt-sock-ro': No such file or directory (Libvirt::ConnectionError)
```

I specify the socket in my `VagrantFile`:
`v.socket = "/Users/vkarri/.cache/libvirt/libvirt-sock"`

If I run the corresponding `box.img` (basically a `qcow2` image) using `qemu-system-x86_64` (e.g. below) the VM starts just fine
```
qemu-system-x86_64 -curses -hda ~/.vagrant.d/boxes/sample/0/libvirt/box.img -m 4G -boot d -netdev user,id=n1,hostfwd=tcp::2222-:22 -device virtio-net-pci,netdev=n1
```

Any ideas?

**Reply by electrofelix** [reply]
To try and access the network information when using user sessions in libvirt, the driver code needs to connect to the system instance of libvirt. Typically this attempt a connection to `qemu:///system` and in this case it appears it results in libvirt trying to open a connection to `/usr/local/var/run/libvirt/libvirt-sock-ro`.

Are you setting the `system_uri` config setting via the Vagrantfile by any chance? If not I believe it must be converting the default of `qemu:///system` to the socket path listed when it calls `Libvirt::open_read_only(config.system_uri)` which is calling https://www.libvirt.org/html/libvirt-libvirt-host.html#virConnectOpenReadOnly

If not, can override the socket by specifying a more complete uri for the config setting `qemu:///system?socket=<path-to-correct-socket>` see https://www.libvirt.org/uri.html for more info.
